DXRModules/DeusEx/Classes/DXRBingoCampaign.uc

```diff
@@ -0,0 +1,391 @@
+    dxr.flags.bingoBoardRoll = 0;
+    dxr.flags.bingo_duration = 1;
+    events.CreateBingoBoard();
+    ClearDataVaultImages(player());
+}
+
+function AddBingoEventBlocker(name blockedTag, name bingoFlag) {
+    local #var(prefix)FlagTrigger ft;
+    local Actor blocked;
+    local name glue;
+
+    glue = dxr.flagbase.StringToName(blockedTag $ "_bingoblocked");
+
+    foreach AllActors(class'Actor', blocked, blockedTag) {
+        if (
+            MapExit(blocked) != None ||
+            InterpolateTrigger(blocked) != None ||
+            DataLinkTrigger(blocked) != None ||
+            Dispatcher(blocked) != None ||
+            DeusExMover(blocked) != None
+        ) {
+            blocked.tag = glue;
+        }
+    }
+
+    ft = Spawn(class'#var(prefix)FlagTrigger',, blockedTag);
+    ft.event = glue;
+    ft.flagName = bingoFlag;
+    ft.bTrigger = true;
+    ft.bTriggerOnceOnly = false;
+    ft.bSetFlag = false;
+    ft.SetCollision(false, false, false);
+}
+
+function UpdateCryptDoors()
+{
+    local #var(prefix)FlagTrigger ft;
+
+    if (
+        dxr.localURL == "10_PARIS_CHATEAU" &&
+        dxr.flagbase.GetBool('everettsignal') &&
+        dxr.flagbase.GetBool(BingoFlagM10) &&
+        IsBingoEnd(dxr.dxInfo.missionNumber, dxr.flags.bingo_duration) &&
+        !dxr.flagbase.GetBool('DXRando_CommandosUnhidden')
+    ) {
+        foreach AllActors(class'#var(prefix)FlagTrigger', ft, 'everettsignaldoor') {
+            ft.Trigger(None, None);
+            break;
+        }
+    }
+}
+
+static function bool IsBingoEnd(int missionNumber, int bingo_duration)
+{
+    if (missionNumber > 12) {
+        missionNumber -= 2;
+    } else if (missionNumber > 6) {
+        missionNumber -= 1;
+    }
+    return missionNumber % bingo_duration == 0 || missionNumber == 13;
+}
+
+static function name GetBingoMissionFlag(int missionNumber) {
+    switch (missionNumber) {
+        case 1:
+            return BingoFlagM01;
+        case 2:
+            return BingoFlagM02;
+        case 3:
+            return BingoFlagM03;
+        case 4:
+            return BingoFlagM04;
+        case 5:
+            return BingoFlagM05;
+        case 6:
+            return BingoFlagM06;
+        case 8:
+            return BingoFlagM08;
+        case 9:
+            return BingoFlagM09;
+        case 10:
+            return BingoFlagM10;
+        case 11:
+            return BingoFlagM11;
+        case 12:
+            return BingoFlagM12;
+        case 14:
+            return BingoFlagM14;
+        case 15:
+            return BingoFlagM15;
+    }
+    return '';
+}
+
+function HandleBingo(int numBingos)
+{
+    if (!dxr.flags.IsBingoCampaignMode() || numBingos < dxr.flags.settings.bingo_win) return;
+
+    if (numBingos == dxr.flags.settings.bingo_win) {
+        player().ClientMessage("You have enough bingo lines to proceed!");
+    }
+
+    switch (dxr.dxInfo.missionNumber) {
+        case 1:
+            dxr.flagbase.SetBool(BingoFlagM01, true,, 2);
```

Comment:
I don't understand what you mean. I use those constants elsewhere and want them to be the authority on what the flags are named.

Reply:
then just make a function that accepts the mission number and outputs the flagname?

Reply:
That function is already there. The reason HandleBingo() doesn't use it is that the mission expiration isn't always the next mission. For instance, the M12 flag doesn't expire until M15 because of backtracking.

Reply:
The function that returns the flagname could also return the expiry

Reply:
That's a good idea.